### PR TITLE
Add migration e2e-test

### PR DESCRIPTION
**Pull request by corest**
Towards https://github.com/giantswarm/giantswarm/issues/3893

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -101,6 +107,10 @@ workflows:
           requires:
           - e2eTestBasic
 
+      - e2eTestMigration:
+          requires:
+          - e2eTestMetrics
```
Comment (rossf7) [inline .circleci/config.yml]: This looks strange to me. Usually the e2e tests just require the build step.

File: integration/test/migration/main_test.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+
+	{
+		c := framework.HostConfig{
+			Logger:     l,
+			ClusterID:  "na",
+			VaultToken: "na",
```
Comment (rossf7) [inline integration/test/migration/main_test.go]: Minor but can you make the config like this to align with other tests.

```
			Logger:     l,

			ClusterID:  "n/a",
			VaultToken: "n/a",
```

File: integration/test/migration/main_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+			RestConfig:      h.RestConfig(),
+			TillerNamespace: "giantswarm",
+		}
+		helmClient, err = helmclient.New(c)
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+
+	{
```
Comment (rossf7) [inline integration/test/migration/main_test.go]: This block should be removed.

The managed services test is only needed for the basic test. See https://github.com/giantswarm/kubernetes-node-exporter/blob/b7aec49844d9b6797fa647ab8758591729cf97f0/integration/test/migration/main_test.go for the wiring for a migration test.

File: integration/test/migration/main_test.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+func TestMain(m *testing.M) {
+	ctx := context.Background()
+
+	{
+		c := e2esetup.Config{
+			HelmClient: helmClient,
+			Host:       h,
+		}
+
+		_, err := e2esetup.Setup(ctx, m, c)
```
Comment (rossf7) [inline integration/test/migration/main_test.go]: This isn't reliable as the tests can fail without an error being returned. Its better to get the exit code from e2esetup and return that with `os.Exit`.

File: integration/test/migration/migration_test.go [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+func checkLegacyResourcesPresent() error {
+	var err error
+
+	c := h.K8sClient()
+	ac := h.K8sAggregationClient()
+	getOptions := metav1.GetOptions{}
+
+	_, err = c.Rbac().ClusterRoleBindings().Get("metrics-server:system:auth-delegator", getOptions)
+	if err != nil {
+		return microerror.Newf("failed to get clusterrolebinding %s: %v", "metrics-server:system:auth-delegator", err)
```
Comment (rossf7) [inline integration/test/migration/migration_test.go]: Not a blocker but current state of the art in tests is to use `%#q` for names. Output is quoted so its easier to read. The #q makes it more reliable but I forget why!

File: integration/test/migration/migration_test.go [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+	listOptions := metav1.ListOptions{
+		LabelSelector: labelSelector,
+	}
+
+	crba, err := c.Rbac().ClusterRoleBindings().List(listOptions)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	if len(crba.Items) != 2 {
+		return microerror.Newf("unexpected number of deployments, want 2, got %d", len(crba.Items))
```
Comment (rossf7) [inline integration/test/migration/migration_test.go]: Deployments?

File: integration/test/migration/migration_test.go [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	l.LogCtx(ctx, "level", "debug", "message", "successfully retrieved metrics from metrics server")
+
+	return nil
+}
+
+func checkLegacyResourcesPresent() error {
```
Comment (rossf7) [inline integration/test/migration/migration_test.go]: Having separate functions for legacy and new resources makes this big. Is it possibly to set the `kind` labels we use in other tests?

https://github.com/giantswarm/kubernetes-kube-state-metrics/blob/afa4767283f4828cd13bc21da34fd3b0b8ad9a41/integration/test/migration/fixtures/resources-chart/templates/resources.yaml#L8

That allows you to use a label selector and just have functions for resources present or not present.

**Review by rossf7 — COMMENTED** [review]
Direction is good here. Just needs a few changes.

File: integration/test/migration/migration_test.go (hunk 7 above)
Comment (corest) [reply integration/test/migration/migration_test.go]: Sure, I wanted to do it in that way. But the issue with it is - there are no labels on the rbac resources, only deployment and service. That's why it is impossible to distinguish those resources from others

File: .circleci/config.yml (hunk 1 above)
Comment (corest) [reply .circleci/config.yml]: Hm, I would say `e2eTestBasic` is required. If helm template is not valid - there is no point to test other things

File: integration/test/migration/migration_test.go (hunk 7 above)
Comment (rossf7) [reply integration/test/migration/migration_test.go]: For the other charts we have a fixtures chart that creates the resources to delete. In that we set the kind label. 

I'm not sure if that works for metrics server. The benefit is it makes the test smaller but as it is now it works so not a blocker.

File: .circleci/config.yml (hunk 1 above)
Comment (rossf7) [reply .circleci/config.yml]: But doesn't that mean the basic test needs to run first? As it is now both tests are run in parallel so you get faster feedback.

IMO the important part is the deployment step requires both e2e tests.

File: integration/test/migration/migration_test.go (hunk 7 above)
Comment (corest) [reply integration/test/migration/migration_test.go]: There is no issue to set those labels in fixture. But in that case test itself doesn't repeat real state of things. I just put in fixtures real resources, which our customers are using

Comment (rossf7) [reply integration/test/migration/migration_test.go]: Yes its true the labels don't exactly match. Although the kind label is ignored by the k8s-migrator job. This was what we ended up with for the previous charts. But if you prefer it this way that's fine.

**Review by rossf7 — APPROVED** [review]
LGTM once CI is green.

**Pull request merged**